In [1]:
import gurobipy as gb
from gurobipy import *
import numpy as np

# Problem 1

In this problem, the aim is to find the optimal stock portfolio that minimizes the portfolio risk subject to non-negative return. 

To do so, Non-Linear Programming will be utilized. 

The performance of different stock options, including their average monthly return and average monthly risk, based on which the optimization will be done, are obtained from historical data.

In [2]:
prob1 = gb.Model("Portfolio - Minimize Risk")

Set parameter Username
Academic license - for non-commercial use only - expires 2023-09-10


In [3]:
stocks = ['FTSE 100', 'DAX', 'DJIA', 'DJ Asian Titans 50', 'Russell 2000']
S = len(stocks)
average_monthly_return = [0.000694877, 0.006401712, 0.006930841, -0.000107527, 0.006875194]
average_monthly_risk = [0.032548849, 0.052860756, 0.032889991, 0.045223062, 0.046334125]
correlation_matrix = [[1, 0.755266863, 0.772833013, 0.687467487, 0.649829211], 
              [0.755266863, 1, 0.703816057, 0.668049744, 0.694259257], 
             [0.772833013, 0.703816057, 1, 0.721454487, 0.8484989], 
             [0.687467487, 0.668049744, 0.721454487, 1, 0.67030134],
             [0.649829211, 0.694259257, 0.8484989, 0.67030134, 1]]

In [4]:
X = prob1.addVars(S, lb = 0, vtype = GRB.CONTINUOUS, name = ["share of "+i for i in stocks])
X

{0: <gurobi.Var *Awaiting Model Update*>,
 1: <gurobi.Var *Awaiting Model Update*>,
 2: <gurobi.Var *Awaiting Model Update*>,
 3: <gurobi.Var *Awaiting Model Update*>,
 4: <gurobi.Var *Awaiting Model Update*>}

In [5]:
# Defining the objective function to minimize risk
prob1.setObjective(sum(X[i]*average_monthly_risk[i]*X[j]*average_monthly_risk[j]*correlation_matrix[i][j] for i in range(S) for j in range(S)), GRB.MINIMIZE)

In [6]:
# Defining the budget constraint
prob1.addConstr(sum(X[i] for i in range(S)) == 1)

<gurobi.Constr *Awaiting Model Update*>

In [7]:
# Defining non-negative overall return contraint
prob1.addConstr(sum(X[i]*average_monthly_return[i] for i in range(S)) >= 0)

<gurobi.Constr *Awaiting Model Update*>

In [8]:
prob1.optimize()

Gurobi Optimizer version 9.5.2 build v9.5.2rc0 (win64)
Thread count: 8 physical cores, 16 logical processors, using up to 16 threads
Optimize a model with 2 rows, 5 columns and 10 nonzeros
Model fingerprint: 0x5763c502
Model has 15 quadratic objective terms
Coefficient statistics:
  Matrix range     [1e-04, 1e+00]
  Objective range  [0e+00, 0e+00]
  QObjective range [2e-03, 7e-03]
  Bounds range     [0e+00, 0e+00]
  RHS range        [1e+00, 1e+00]
Presolve time: 0.03s
Presolved: 2 rows, 5 columns, 10 nonzeros
Presolved model has 15 quadratic objective terms
Ordering time: 0.00s

Barrier statistics:
 Free vars  : 4
 AA' NZ     : 1.500e+01
 Factor NZ  : 2.100e+01
 Factor Ops : 9.100e+01 (less than 1 second per iteration)
 Threads    : 1

                  Objective                Residual
Iter       Primal          Dual         Primal    Dual     Compl     Time
   0   2.47487651e+05 -2.47487651e+05  5.00e+03 1.07e-06  1.00e+06     0s
   1   4.06845580e+03 -4.14932416e+03  3.51e+02 7.52e-

In [10]:
print(np.sqrt(prob1.ObjVal))
for v in prob1.getVars():
    print(v.varName, "=", v.x)

0.030801131246480293
share of FTSE 100 = 0.5229435905708266
share of DAX = 4.019429819458159e-09
share of DJIA = 0.477056278332486
share of DJ Asian Titans 50 = 1.1694744401191377e-07
share of Russell 2000 = 1.0129816815051385e-08


According to the results, 52 percent of the available funds should be invested in FTSE 100 and the remaining 48 percent should be invested in DJIA to minimize the risk of the portfolio according to the historical performance of the studies stocks.

# Problem 1b

Now we assume there is already an existing portfolio and try to optimize it. Currently, 20 percent of the available funds is invested in each of the 5 stock options. Changing the allocation requires incurring transaction costs.

Similar to the previous case, the aim is to minimize the risk of the portfolio.

Contrary to the previous case, we are not starting to invest in this example. Rather, we want to improve the performance of an existing portfolio by minimizing its risk.

Again, Non-Linear Programming will be used to solve the problem. 

In [11]:
prob1b = gb.Model("Portfolio - Minimize Risk - Initial Allocation")

In [12]:
stocks = ['FTSE 100', 'DAX', 'DJIA', 'DJ Asian Titans 50', 'Russell 2000']
S = len(stocks)
average_monthly_return = [0.000694877, 0.006401712, 0.006930841, -0.000107527, 0.006875194]
average_monthly_risk = [0.032548849, 0.052860756, 0.032889991, 0.045223062, 0.046334125]
correlation_matrix = [[1, 0.755266863, 0.772833013, 0.687467487, 0.649829211], 
              [0.755266863, 1, 0.703816057, 0.668049744, 0.694259257], 
             [0.772833013, 0.703816057, 1, 0.721454487, 0.8484989], 
             [0.687467487, 0.668049744, 0.721454487, 1, 0.67030134],
             [0.649829211, 0.694259257, 0.8484989, 0.67030134, 1]]
initial_allocation = [0.2, 0.2, 0.2, 0.2, 0.2]

In [13]:
X = prob1b.addVars(S, lb = 0, vtype = GRB.CONTINUOUS, name = ["share of "+i for i in stocks])
X

{0: <gurobi.Var *Awaiting Model Update*>,
 1: <gurobi.Var *Awaiting Model Update*>,
 2: <gurobi.Var *Awaiting Model Update*>,
 3: <gurobi.Var *Awaiting Model Update*>,
 4: <gurobi.Var *Awaiting Model Update*>}

In [14]:
# Defining the objective function to minimize risk
prob1b.setObjective(sum(X[i]*average_monthly_risk[i]*X[j]*average_monthly_risk[j]*correlation_matrix[i][j] for i in range(S) for j in range(S)), GRB.MINIMIZE)

In [15]:
# Defining the budget constraint
prob1b.addConstr(sum(X[i] for i in range(S)) == 1)

<gurobi.Constr *Awaiting Model Update*>

In [16]:
# Defining new variables to model absolute function in the upcoming return constraint
D = prob1b.addVars(S, vtype = GRB.CONTINUOUS, name = ["change in share of "+i for i in stocks])
D

{0: <gurobi.Var *Awaiting Model Update*>,
 1: <gurobi.Var *Awaiting Model Update*>,
 2: <gurobi.Var *Awaiting Model Update*>,
 3: <gurobi.Var *Awaiting Model Update*>,
 4: <gurobi.Var *Awaiting Model Update*>}

In this problem, we should also consider the transaction cost since we have an initial allocation 

Therefore, we would subtract the transaction cost from the overall return

In [17]:
# Defining non-negative overall return contraint 

prob1b.addConstr(sum(X[i]*average_monthly_return[i] for i in range(S)) - sum(D[i] for i in range(S)) >= 0)

<gurobi.Constr *Awaiting Model Update*>

In [18]:
for i in range(S):
    prob1b.addConstr(np.power((X[i] - initial_allocation[i]), 2) == D[i])

In [19]:
# setting non-convex parameter
prob1b.params.NonConvex = 2

Set parameter NonConvex to value 2


In [20]:
prob1b.optimize()

Gurobi Optimizer version 9.5.2 build v9.5.2rc0 (win64)
Thread count: 8 physical cores, 16 logical processors, using up to 16 threads
Optimize a model with 2 rows, 10 columns and 15 nonzeros
Model fingerprint: 0xc16ecc77
Model has 15 quadratic objective terms
Model has 5 quadratic constraints
Coefficient statistics:
  Matrix range     [1e-04, 1e+00]
  QMatrix range    [1e+00, 1e+00]
  QLMatrix range   [4e-01, 1e+00]
  Objective range  [0e+00, 0e+00]
  QObjective range [2e-03, 7e-03]
  Bounds range     [0e+00, 0e+00]
  RHS range        [1e+00, 1e+00]
  QRHS range       [4e-02, 4e-02]

Continuous model is non-convex -- solving as a MIP

Presolve time: 0.01s
Presolved: 17 rows, 10 columns, 40 nonzeros
Presolved model has 15 quadratic objective terms
Presolved model has 5 bilinear constraint(s)
Variable types: 10 continuous, 0 integer (0 binary)

Root relaxation: objective 1.260001e-03, 27 iterations, 0.00 seconds (0.00 work units)

    Nodes    |    Current Node    |     Objective Bounds  

In [21]:
print(np.sqrt(prob1b.ObjVal))
for v in prob1b.getVars():
    print(v.varName, "=", v.x)

0.03588590942740207
share of FTSE 100 = 0.23520946331820236
share of DAX = 0.1608507656394297
share of DJIA = 0.2301438086960881
share of DJ Asian Titans 50 = 0.18945448905852297
share of Russell 2000 = 0.18434147328775682
change in share of FTSE 100 = 0.0012392801663399025
change in share of DAX = 0.001532052959102746
change in share of DJIA = 0.0009081677204066853
change in share of DJ Asian Titans 50 = 0.00011119001710135044
change in share of Russell 2000 = 0.00024457312335575843


According to the results, the optimal allocation would be as follows:

FTSE 100 = 0.235 (+0.035)

DAX = 0.161 (-0.039)

DJIA = 0.23 (+0.03)

DJ Asian Titans 50 = 0.189 (-0.011)

DJ Asian Titans 50 = 0.185 (-0.015)

# Problem 2

Now, we want to optimize resource allocation for an R&D department. There are 6 projects available but only 25 employees to work on them. 

Each project requires an initial investment.

The probability of success for each project is a non-linear function of the number of individuals working on the project.

The expected profit of each project in case of success in also known. 

Engineers can also work on multiple projects meaning that, for example, it is possible for an employee to dedicate 20 percent of his/her time to one project and 80 percent of it on another one. 

Also, it is worth mentioning that the company does not necessarily have to assign all employees to a project.

Non-Linear Programming is utilized to find the optimal allocation of employees to maximize the total profit.

In [2]:
prob2 = gb.Model("Allocation - Maximize Profit")

Set parameter Username
Academic license - for non-commercial use only - expires 2023-09-10


In [3]:
cost = [325000, 200000, 490000, 125000, 710000, 240000]
profit = [1750000, 700000, 1300000, 800000, 1450000, 1300000]
constant = [1.1, 0.5, 2.5, 1.6, 2.2, 2.4]
N = len(profit)

In [4]:
X = prob2.addVars(N, lb = 0, vtype = GRB.CONTINUOUS, name = ["Engineers for project "+str(i) for i in range(N)])
X

{0: <gurobi.Var *Awaiting Model Update*>,
 1: <gurobi.Var *Awaiting Model Update*>,
 2: <gurobi.Var *Awaiting Model Update*>,
 3: <gurobi.Var *Awaiting Model Update*>,
 4: <gurobi.Var *Awaiting Model Update*>,
 5: <gurobi.Var *Awaiting Model Update*>}

In [5]:
# Add new variables to linearize the objective function / model variables in denominators in a different way
Z = prob2.addVars(N, vtype = GRB.CONTINUOUS, name = ["Denominator "+str(i) for i in range(N)])
Z

{0: <gurobi.Var *Awaiting Model Update*>,
 1: <gurobi.Var *Awaiting Model Update*>,
 2: <gurobi.Var *Awaiting Model Update*>,
 3: <gurobi.Var *Awaiting Model Update*>,
 4: <gurobi.Var *Awaiting Model Update*>,
 5: <gurobi.Var *Awaiting Model Update*>}

The objective funtion would be the profit for each project multiplied by its success probability subracted by the cost for each project multiplied by its failure probability and the total cost of engineers.

In [6]:
# Defining the objective function to maximize profit

prob2.setObjective(sum(X[i]*Z[i]*profit[i] for i in range(N)) - sum(constant[i]*Z[i]*cost[i] for i in range(N)) - 150000*sum(X[i] for i in range(N)), GRB.MAXIMIZE)

In [7]:
for i in range(N):
    prob2.addConstr(Z[i] * (X[i] + constant[i]) == 1)

In [8]:
# setting non-convex parameter
prob2.params.NonConvex = 2

Set parameter NonConvex to value 2


In [9]:
# Defining the constraint for maximum number of engineers
prob2.addConstr(sum(X[i] for i in range(N)) <= 25)

<gurobi.Constr *Awaiting Model Update*>

In [10]:
prob2.optimize()

Gurobi Optimizer version 9.5.2 build v9.5.2rc0 (win64)
Thread count: 8 physical cores, 16 logical processors, using up to 16 threads
Optimize a model with 1 rows, 12 columns and 6 nonzeros
Model fingerprint: 0xb53beea4
Model has 6 quadratic objective terms
Model has 6 quadratic constraints
Coefficient statistics:
  Matrix range     [1e+00, 1e+00]
  QMatrix range    [1e+00, 1e+00]
  QLMatrix range   [5e-01, 3e+00]
  Objective range  [1e+05, 2e+06]
  QObjective range [1e+06, 4e+06]
  Bounds range     [0e+00, 0e+00]
  RHS range        [3e+01, 3e+01]
  QRHS range       [1e+00, 1e+00]

Continuous model is non-convex -- solving as a MIP

Presolve time: 0.01s
Presolved: 34 rows, 17 columns, 79 nonzeros
Presolved model has 10 bilinear constraint(s)
Variable types: 17 continuous, 0 integer (0 binary)

Root relaxation: objective 2.351842e+07, 24 iterations, 0.01 seconds (0.00 work units)

    Nodes    |    Current Node    |     Objective Bounds      |     Work
 Expl Unexpl |  Obj  Depth IntInf |

In [11]:
for v in prob2.getVars():
    print(v.varName, "=", v.x)

Engineers for project 0 = 2.8008151444040212
Engineers for project 1 = 1.2321728299220254
Engineers for project 2 = 2.9626273455106036
Engineers for project 3 = 1.5412394702310088
Engineers for project 4 = 3.4284538698771407
Engineers for project 5 = 2.5635513065961684
Denominator 0 = 0.2563566749464061
Denominator 1 = 0.5773093033778878
Denominator 2 = 0.18306209388817976
Denominator 3 = 0.3183453736139061
Denominator 4 = 0.17766868541854605
Denominator 5 = 0.20146865383885107


According to the results, the optimal allocation would be as follows:

Number of employees working on project 0 = 2.80

Number of employees working on project 1 = 1.23

Number of employees working on project 2 = 2.96

Number of employees working on project 3 = 1.54

Number of employees working on project 4 = 3.43

Number of employees working on project 5 = 2.56

It is worth mentioning that, for example, 2.56 employees working on project 5 means that the amount of effort put on the project is equal to the amount of work that 2.56 full time employees do with any combination. 

# Problem 2b

Now, due to a change in business requirements, we optimize the allocation to minimize the standard deviation of the contribution to profit subject to the constraint that the expected contribution to profit (minus the cost of employees) is at least $1.1 million.

The rest is similar to Problem 2.

In [2]:
prob2b = gb.Model("Allocation - Minimize SD")

Set parameter Username
Academic license - for non-commercial use only - expires 2023-09-10


In [3]:
cost = [325, 200, 490, 125, 710, 240]
profit = [1750, 700, 1300, 800, 1450, 1300]
constant = [1.1, 0.5, 2.5, 1.6, 2.2, 2.4]
N = len(profit)

In [4]:
X = prob2b.addVars(N, lb = 0, vtype = GRB.CONTINUOUS, name = ["Engineers for project "+str(i) for i in range(N)])
X

{0: <gurobi.Var *Awaiting Model Update*>,
 1: <gurobi.Var *Awaiting Model Update*>,
 2: <gurobi.Var *Awaiting Model Update*>,
 3: <gurobi.Var *Awaiting Model Update*>,
 4: <gurobi.Var *Awaiting Model Update*>,
 5: <gurobi.Var *Awaiting Model Update*>}

In [5]:
S = prob2b.addVars(N, vtype = GRB.CONTINUOUS, name = ["Success Probability "+str(i) for i in range(N)])
S

{0: <gurobi.Var *Awaiting Model Update*>,
 1: <gurobi.Var *Awaiting Model Update*>,
 2: <gurobi.Var *Awaiting Model Update*>,
 3: <gurobi.Var *Awaiting Model Update*>,
 4: <gurobi.Var *Awaiting Model Update*>,
 5: <gurobi.Var *Awaiting Model Update*>}

In [6]:
# Defining the objective function to minimize the standard deviation
prob2b.setObjective(sum(S[i]*(1 - S[i])*np.power((profit[i] + cost[i]), 2) for i in range(N)), GRB.MINIMIZE)

In [7]:
for i in range(N):
    prob2b.addConstr(S[i] * (X[i] + constant[i]) == X[i])

In [8]:
# Defining the minimum return constraint
# Return should be more than $1.1 million
# Return consists of the outcome of the project (success + and failure -) subtracted by the cost of engineers
prob2b.addConstr(sum(S[i]*profit[i] for i in range(N)) - sum((1 - S[i])*cost[i] for i in range(N)) - 150*sum(X[i] for i in range(N)) >= 1100)

<gurobi.Constr *Awaiting Model Update*>

In [9]:
# Defining the constraint for maximum number of engineers 
prob2b.addConstr(sum(X[i] for i in range(N)) <= 25)

<gurobi.Constr *Awaiting Model Update*>

In [10]:
# Setting non-convex parameter 
prob2b.params.NonConvex = 2

Set parameter NonConvex to value 2


In [11]:
prob2b.optimize()

Gurobi Optimizer version 9.5.2 build v9.5.2rc0 (win64)
Thread count: 8 physical cores, 16 logical processors, using up to 16 threads
Optimize a model with 2 rows, 12 columns and 18 nonzeros
Model fingerprint: 0x1bfab73e
Model has 6 quadratic objective terms
Model has 6 quadratic constraints
Coefficient statistics:
  Matrix range     [1e+00, 2e+03]
  QMatrix range    [1e+00, 1e+00]
  QLMatrix range   [5e-01, 3e+00]
  Objective range  [8e+05, 5e+06]
  QObjective range [2e+06, 9e+06]
  Bounds range     [0e+00, 0e+00]
  RHS range        [3e+01, 3e+03]

Continuous model is non-convex -- solving as a MIP

Presolve time: 0.00s
Presolved: 33 rows, 19 columns, 85 nonzeros
Presolved model has 12 bilinear constraint(s)
Variable types: 19 continuous, 0 integer (0 binary)

Root relaxation: objective -1.964960e+09, 18 iterations, 0.00 seconds (0.00 work units)

    Nodes    |    Current Node    |     Objective Bounds      |     Work
 Expl Unexpl |  Obj  Depth IntInf | Incumbent    BestBd   Gap | It/

In [12]:
for v in prob2b.getVars():
    print(v.varName, "=", v.x)

Engineers for project 0 = 5.051188581982795
Engineers for project 1 = 1.6467656216754682
Engineers for project 2 = 4.208168532917417
Engineers for project 3 = 1.4739860340122422
Engineers for project 4 = 5.9830772013212075
Engineers for project 5 = 3.046289721024369
Success Probability 0 = 0.8211727724911626
Success Probability 1 = 0.7670914817381093
Success Probability 2 = 0.6273200370962152
Success Probability 3 = 0.4795031655001891
Success Probability 4 = 0.7311524814106853
Success Probability 5 = 0.5593330279997307


According to the results, the optimal allocation would be as follows:

Number of employees working on project 0 = 5.05

Number of employees working on project 1 = 1.65

Number of employees working on project 2 = 4.21

Number of employees working on project 3 = 1.47

Number of employees working on project 4 = 5.98

Number of employees working on project 5 = 3.05

Similar to the previous problem (Problem 2), for example, 3.05 employees working on project 5 means that the effort put on project 5 is equal to the work of 5.05 full time employees. 

It can be seen that utilization ratio in this case is higher than the previous example (85.6% vs 58.1%)